### PR TITLE
[security, licenses] Document, partially automate, 3PP license and vulnerability checks

### PR DESCRIPTION
**Pull request by marcdumais-work**
Add README-3PP-LICENSE-CHECK-AND-AUDIT.md to document how to perform
license and vulnerability checks for vscode builtin extensions (internal
and external).

Add script that clones and checks-out vscode external builtin repos.

src/get-external-builtins.js

When run, each external builtin repo is cloned and then checked-out,
as per its definition in vscode's `product.json` file (section
BuiltInExtensions).

For convenience, also added a yarn script to run this:
$> yarn get-external-builtins

### CODE REVIEW

**Comment by CiciDongxue** [comment]
您好，我已收到您的来信，暂时无法回复您的邮件，望见谅。

File: src/get-external-builtins.js
Comment (alvsan09) [inline src/get-external-builtins.js:72]: I would prefer not to catch errors and propagate them up to abort the flow and have a single top try/catch location where we report the error.

What do you think ?

File: src/get-external-builtins.js
Comment (alvsan09) [inline src/get-external-builtins.js:39]: remove the ending `and` ?

File: src/get-external-builtins.js
Comment (alvsan09) [inline src/get-external-builtins.js:37]: shouldn't we cleanup the `externalBuiltinsRepos` directory before entering the loop?
otherwise, we may end up preserving an old version of the plugin

File: src/get-external-builtins.js
Comment (alvsan09) [inline src/get-external-builtins.js:42]: if we cleanup the directory `externalBuiltinsRepos` before entering the loop, we don't need to check if a plugin folder already exists, and if does, it means that the production.json has an errors and it's have duplicate entries, which is unlikely and not worth covering for it. 
any thoughts ?

**Review by alvsan09 — CHANGES_REQUESTED** [review]
Very interesting functionality which needs more automation in later changes as you have highlighted !! 
See my inline comments !

File: README-3PP-LICENSE-CHECK-AND-AUDIT.md
Comment (marcdumais-work) [reply README-3PP-LICENSE-CHECK-AND-AUDIT.md:20]: Since this is required to do for all checks documented in this procedure, I have added this information towards the beginning of the document.

File: README-3PP-LICENSE-CHECK-AND-AUDIT.md
Comment (marcdumais-work) [reply README-3PP-LICENSE-CHECK-AND-AUDIT.md:24]: Good question: yes it is. The EF license scan tool, `scancode` does that. For example, see the report attached to the issue I raised on Gitlab for the "internal" builtins IP check:
https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/4120#note_1031903

I have submitted, for the same analysis, the external builtins that we currently use, individually (since they each have their own repo, they are easy to process this way automatically):
https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/4121
https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/4122
https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/4123
https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/4124

File: README-3PP-LICENSE-CHECK-AND-AUDIT.md
Comment (marcdumais-work) [reply README-3PP-LICENSE-CHECK-AND-AUDIT.md:23]: Good catch. I will come-back to this one after taking care of the easier-to-fix feedback.

File: README-3PP-LICENSE-CHECK-AND-AUDIT.md
Comment (marcdumais-work) [reply README-3PP-LICENSE-CHECK-AND-AUDIT.md:44]: Fixed in both places

File: README-3PP-LICENSE-CHECK-AND-AUDIT.md
Comment (marcdumais-work) [reply README-3PP-LICENSE-CHECK-AND-AUDIT.md:52]: I removed the specific example. FYI, I noticed that the number of `external builtins` looks to go down as we progress in vscode versions, and they tend to be available on `open vsx`. Momentarily (1.55.2) we are still missing a few minor ones.

File: README-3PP-LICENSE-CHECK-AND-AUDIT.md
Comment (marcdumais-work) [reply README-3PP-LICENSE-CHECK-AND-AUDIT.md:56]: You're right - the `.js` script is a node script. Here, I meant the added entry in the `scripts` section of `package.json`,  that invokes the `node` script. I changed "node" to "package.json" so it's clear.

File: README-3PP-LICENSE-CHECK-AND-AUDIT.md
Comment (marcdumais-work) [reply README-3PP-LICENSE-CHECK-AND-AUDIT.md:65]: I did that on purpose so that we do not need to check for its existence or create the directory. `external-builtin-extensions-here.txt` is a place-holder, that can be removed if we eventually have real content under that directory. (at first this prodecure was in there, but I moved it up to the root when I generalized it to both internal and external builtins).

Does it make sense?

File: README-3PP-LICENSE-CHECK-AND-AUDIT.md
Comment (marcdumais-work) [reply README-3PP-LICENSE-CHECK-AND-AUDIT.md:66]: Correct - it does not. ATM, there is a single extension in that section, and its [repo](https://github.com/microsoft/vscode-remotehub) no longer exists (404). 

More generally, MS started to include the potential to build "web" versions of some internal builtins in the repo's infrastructure (build scripts), but this seems to be WIP at 1.55.2. 
e.g.:
from `<vscode>/built/lib/extensions.ts`:
image

We will have another look when we progress to more recent baselines.

File: src/get-external-builtins.js
Comment (marcdumais-work) [reply src/get-external-builtins.js:37]: The presence of old runs should not interfere with the script, even if we switch `vscode` baseline along the way. 

That being said, it's generally a good thing to start from a known, clean state. So, I have added a `yarn clean` instruction (that calls `git clean -ffdx`) towards the top of the `README-3PP-LICENNSE-CHECK_AND_AUDIT.md`, that will clear the external builtin repos among other things.

File: src/get-external-builtins.js
Comment (marcdumais-work) [reply src/get-external-builtins.js:72]: I have moved the try-catch blocks one level up as a single one. I do not abort the flow currently, but I added a summary printout at the end of execution, to make it easier to notice that any given step (git clone or checkout) has failed.

File: README-3PP-LICENSE-CHECK-AND-AUDIT.md
Comment (marcdumais-work) [reply README-3PP-LICENSE-CHECK-AND-AUDIT.md:23]: I have sub-projects covered for the internal built-ins. It found one extra dependency that needed approval : https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/4694

Comment (marcdumais-work) [reply README-3PP-LICENSE-CHECK-AND-AUDIT.md:23]: I did a similar change for external built-ins. We should now cover better sub-projects (deeper that repo root) for both types of built-ins.

Comment (marcdumais-work) [reply README-3PP-LICENSE-CHECK-AND-AUDIT.md:23]: This allowed to find a few more dependencies to vet, for external  built-ins:
image

File: README-3PP-LICENSE-CHECK-AND-AUDIT.md
Comment (marcdumais-work) [reply README-3PP-LICENSE-CHECK-AND-AUDIT.md:47]: sub-packages are now covered

File: src/paths.js
Comment (marcdumais-work) [reply src/paths.js:61]: I have done as you suggest here and removed the folder from source control :)

File: README-3PP-LICENSE-CHECK-AND-AUDIT.md
Comment (marcdumais-work) [reply README-3PP-LICENSE-CHECK-AND-AUDIT.md:23]: FYI, instead of relying on a folder depth, the new way relies on finding `yarn.lock` and/or `package-lock.json` files, no matter the depth. The tools we use (yarn/npm audit, dash-licenses) require one of these to work, so I think we cover as much as possible with this strategy.

File: README-3PP-LICENSE-CHECK-AND-AUDIT.md
Comment (marcdumais-work) [reply README-3PP-LICENSE-CHECK-AND-AUDIT.md:65]: update: I have done as suggested below and removed this file and folder from source control.

File: src/get-external-builtins.js
Comment (marcdumais-work) [reply src/get-external-builtins.js:46]: True. For now I have not implemented that suggested change, in favor of documenting how to start the process with a clean slate. I think this gives more flexibility, if we need to re-run the script, e.g. to re-clone one repo or checkout version corresponding to another `vscode` baseline.

**Comment by marcdumais-work** [comment]
Thanks for the review @alvsan09 . I have done most of the suggested changes. We can sync-up and discuss further if you want.

**Review by alvsan09 — COMMENTED** [review]
I have tested the commands using dash-licenses without the `review` parameter and they work as expected :+1: 

I just saw few cosmetics that are marked as inline comments.
It would be good to have second pair of eyes on this change so I will invite @vince-fugnitto 

Thanks Marc !!

File: README-3PP-LICENSE-CHECK-AND-AUDIT.md
Comment (marcdumais-work) [reply README-3PP-LICENSE-CHECK-AND-AUDIT.md:38]: > I assume you are updating the following issue accordingly?: https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/4120#note_1031903

No - I added the info of what was found initially to show we would also cover the 3PP dependencies by `dash-licenses` on those extensions, but there is no real purpose in keeping an updated list. We do not keep such a list for the main repo either. 

If needed we have the means to perform a "reverse mapping" of sorts by re-running `dash-licenses` on this version of the vscode-builtins and confirming that no dependency is flagged as suspicious by `dash-license` (except `works-with` approved for us, but not yet correctly handled by the tool, like `jschardet`)

**Comment by marcdumais-work** [comment]
Thanks @alvsan09 - I have pushed a new commit to address the newer feedback. (I will squash before merging)

I have also found an issue with some of the commands that would be too eager at finding (yarn/npm) lock files under `node_modules` and then running license and vulnerability checks for these, finding extra dependencies that should be ignored (e.g. deep dev-dependencies). 

In consequence, we now ignore lock files that have `node_modules` in their relative path.

**Review by alvsan09 — APPROVED** [review]
It LGTM as well !!
Thanks Marc

**Pull request merged**
